src/lib.rs

```diff
@@ -588,6 +595,182 @@ fn floyd_warshall(py: Python, dag: &digraph::PyDiGraph) -> PyResult<PyObject> {
     Ok(out_dict.into())
 }
 
+fn get_edge_iter_with_weights<G>(
+    graph: G,
+) -> impl Iterator<Item = (usize, usize, PyObject)>
+where
+    G: GraphBase
+        + IntoEdgeReferences
+        + IntoNodeIdentifiers
+        + NodeIndexable
+        + GraphProp
+        + NodesRemoved,
+    G: Data<NodeWeight = PyObject, EdgeWeight = PyObject>,
+{
+    let node_map: Option<HashMap<NodeIndex, usize>>;
+    if graph.nodes_removed() {
+        let mut node_hash_map: HashMap<NodeIndex, usize> = HashMap::new();
+        for (count, node) in graph.node_identifiers().enumerate() {
+            let index = NodeIndex::new(graph.to_index(node));
+            node_hash_map.insert(index, count);
+        }
+        node_map = Some(node_hash_map);
+    } else {
+        node_map = None;
+    }
+
+    graph.edge_references().map(move |edge| {
+        let i: usize;
+        let j: usize;
+        match &node_map {
+            Some(map) => {
+                let source_index =
+                    NodeIndex::new(graph.to_index(edge.source()));
+                let target_index =
+                    NodeIndex::new(graph.to_index(edge.target()));
+                i = *map.get(&source_index).unwrap();
+                j = *map.get(&target_index).unwrap();
+            }
+            None => {
+                i = graph.to_index(edge.source());
+                j = graph.to_index(edge.target());
+            }
+        }
+        (i, j, edge.weight().clone())
```

Comment:
Probably a very basic question, Why do we need `clone()` (or clone_ref) here? What happens if we omit `clone()`?

Reply:
Oh, this actually a really good (and pretty involved) question and is actually the one thing in the PR I'm not happy with, but it was the only way I managed to make it work. The `clone()` is necessary because the return type from this function for weight is of type `PyObject` but `edge.weight()` returns a `&PyObject`. If you remove the `clone()` you get a type mismatch error:

```
error[E0271]: type mismatch resolving `<[closure@src/lib.rs:622:33: 640:6 node_map:_, graph:_] as std::ops::FnOnce<(<G as petgraph::visit::IntoEdgeReferences>::EdgeRef,)>>::Output == (usize, usize, pyo3::PyObject)`
   --> src/lib.rs:600:6
    |
600 | ) -> impl Iterator<Item = (usize, usize, PyObject)>
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&pyo3::PyObject`, found struct `pyo3::PyObject`
    |
    = note: expected tuple `(_, _, &pyo3::PyObject)`
               found tuple `(_, _, pyo3::PyObject)`
    = note: required because of the requirements on the impl of `std::iter::Iterator` for `std::iter::Map<<G as petgraph::visit::IntoEdgeReferences>::EdgeReferences, [closure@src/lib.rs:622:33: 640:6 node_map:_, graph:_]>`
    = note: the return type of a function must have a statically known size

error: aborting due to previous error

For more information about this error, try `rustc --explain E0271`.
error: could not compile `retworkx`.
```

There are 2 paths to fix the type mismatch besides a `clone()` here, either dereference the borrow or change the return type to a reference. The approach which I actually wanted to do at first was change the return type to a reference. But the problem here is the [lifetime](https://doc.rust-lang.org/1.9.0/book/lifetimes.html) of the reference. The borrow checker doesn't know how long the reference will live which results in an error:

```error[E0106]: missing lifetime specifier
   --> src/lib.rs:600:42
    |
600 | ) -> impl Iterator<Item = (usize, usize, &PyObject)>
    |                                          ^ expected named lifetime parameter
    |
    = help: this function's return type contains a borrowed value with an elided lifetime, but the lifetime cannot be derived from the arguments
help: consider using the `'static` lifetime
    |
600 | ) -> impl Iterator<Item = (usize, usize, &'static PyObject)>
    |                                          ^^^^^^^^

error[E0310]: the associated type `<G as petgraph::visit::IntoEdgeReferences>::EdgeRef` may not live long enough
   --> src/lib.rs:639:16
    |
639 |         (i, j, edge.weight())
    |                ^^^^
    |
    = help: consider adding an explicit lifetime bound `<G as petgraph::visit::IntoEdgeReferences>::EdgeRef: 'static`...
    = note: ...so that the type `<G as petgraph::visit::IntoEdgeReferences>::EdgeRef` is not borrowed for too long

error[E0310]: the associated type `<G as petgraph::visit::IntoEdgeReferences>::EdgeRef` may not live long enough
   --> src/lib.rs:639:21
    |
639 |         (i, j, edge.weight())
    |                     ^^^^^^
    |
    = help: consider adding an explicit lifetime bound `<G as petgraph::visit::IntoEdgeReferences>::EdgeRef: 'static`...
    = note: ...so that the reference type `&<G as petgraph::visit::IntoEdgeReferences>::EdgeRef` does not outlive the data it points at

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0106, E0310.
For more information about an error, try `rustc --explain E0106`.
```

Normally you can fix this by setting an explicit lifetime to tell the compiler how long the reference will live so it can ensure that the operations on the reference are safe or not. But, in this case the compiler error message is pointing out that the lifetime of the reference is defined in the trait definition for `IntoEdgeReferences`, which already has a lifetime set on it. I wasn't able to find a way to specify that the lifetime of the returned `PyObject` reference from the iterator was the same as the lifetime for the `PyGraph` object (which is specified here: https://github.com/Qiskit/retworkx/blob/master/src/graph.rs#L132 ). I think this is why the compiler is suggestion that I use a [static lifetime](https://doc.rust-lang.org/stable/rust-by-example/scope/lifetime/static_lifetime.html#reference-lifetime) for the returned `&PyObject` even though that wouldn't work (the edge reference doesn't have a static lifetime).

So without a way to pass a reference we can look at dereferencing the return from `edge.weight()`. The problem is if we dereference the borrow like `*edge.weight()` the lifetime of the borrow will expire when the function exits and the value will be dropped. The compiler errors telling us this with:

```
error[E0507]: cannot move out of a shared reference
   --> src/lib.rs:639:16
    |
639 |         (i, j, *edge.weight())
    |                ^^^^^^^^^^^^^^ move occurs because value has type `pyo3::PyObject`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0507`.
error: could not compile `retworkx`.
```

Using `clone()` here is just how we can avoid that data ownership issue. By returning a "copy" of the data instead of value from a reference. However, `clone()` of a PyObject isn't actually a memory copy it just increases the refcount in python: https://docs.rs/pyo3/0.11.1/src/pyo3/object.rs.html#311-318 so it typically shouldn't be that slow. It just needs a GIL context to call python to do that though, which will be the bottleneck on `clone()`.

It's also worth pointing out that this wasn't an issue in the duplicated version because the scope (each of the functions) was the same so we didn't have to worry about the lifetime of the data.

Reply:
Thank you for your detailed answer. I understand I have to learn more about lifetime and ownership:-) I've played with tweaking the code a bit and getting error messages like you showed above, and found there is no other way than using `clone()` since we have no means to change the lifetime of `EdgeRef` defined in the `petgraph` library. So now `clone()` makes sense to me.

Reply:
FWIW, borrowing, lifettimes, etc are personally what I find to be the hardest to get used to in Rust. I've personally only recently felt like I'm comfortable dealing with the concepts relatively recently. When I first started working in rust I felt like I was just brute forcing my way through compiler errors. But eventually after working with the language for a while I started to understand more of it. The compiler messages and `rustc --explain` help a lot normally actually. It's just in this case petgraph (and some other libs) just make the error messages harder to read, because they're written to be super generic the error messages don't make a lot of sense without understanding how petgraph works.